### ISSUE TITLE
Koodikatselmointi #2

### ISSUE THREAD

**Issue by nkoiv** [Salmela/particles#2]
(project cloned at 2015/10/08 21:00)

**Project structure**
Reading the general description and specs didnt really open up the project all too well - at least not to someone as uneducated as me. I got the general idea of what's being done, but how it's to be done was quite confusing. You could perhaps consider adding in a class diagram to visualize the project structure, as well as maybe do some sequence diagrams on the key aspects of the program.
Packaging classes in a structured way (instead of having everything under org.tiralab.particles) would also help when delving into the program code the for the first time. Going into main class and diving deeper from there works, but it takes a while to understand what's governing what. I tried building the javadoc to read a bit from, but it was also slightly holey, if you will. Some classes (ParticleArray for example) were very well covered, while others (Stars, Fluids, DirectStorage... ) less so.

**Code clarity**
Code itself is easy to read and nicely condensed. There are a few long methods (simulate() at Stars and Fluid in paricular), but I at least to me the added comments within helped a lot with the readability. For the most of it, the code needed no comments - good expressive method names and variables made it a breeze to understand basic functions.

**Bugs**
Adding in some particles and then ressing 'A' a couple of times caused a stackoverflow-error (at org.tiralab.particles.TreeStorage$QuadTreeNode.<init>(TreeStorage.java:166)), but beyond that the program seemed mostly stable. This seemed to be reproducable by going from TreeStorage in particular. Perhaps a problem with the storage not extending as it should when enough stuff is added fast enough?

**TLDR;**
Clean code, nice project. Could use some more documenting.
